### PR TITLE
fix: Avoid controller crashes when running large number of workflows

### PR DESCRIPTION
**Pull request by terrytangyuan**
Fixes https://github.com/argoproj/argo-workflows/issues/8275.

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>

### CODE REVIEW

File: workflow/controller/healthz.go [hunk 1]
```diff
@@ -30,7 +30,8 @@ func (wfc *WorkflowController) Healthz(w http.ResponseWriter, r *http.Request) {
 	labelSelector := "!" + common.LabelKeyPhase + "," + instanceIDSelector
 	err := func() error {
 		// avoid problems with informers, but directly querying the API
-		list, err := wfc.wfclientset.ArgoprojV1alpha1().Workflows(wfc.managedNamespace).List(ctx, metav1.ListOptions{LabelSelector: labelSelector})
+		limit := int64(env.LookupEnvIntOr("LIST_LIMIT", 200))
```
Comment (jessesuen) [inline workflow/controller/healthz.go:33]: Lets call this `HEALTHZ_LIST_LIMIT` since it only applies to healthz.

Comment (terrytangyuan) [reply workflow/controller/healthz.go]: Great suggestion. Updated.

File: docs/environment-variables.md [hunk 2]
```diff
@@ -33,12 +33,14 @@ most users. Environment variables may be removed at any time.
...
 | `LEADER_ELECTION_RETRY_PERIOD`         | `time.Duration`     | `5s`                                                                                        | The duration that the leader election clients should wait between tries of actions.                                                                                                                                                                                      |
+| `LIST_LIMIT`                           | `int`               | `200`                                                                                       | The maximum number of responses to return for a list call on workflows for workflow informer.                                                                                                                                                                            |
```
Comment (jessesuen) [inline docs/environment-variables.md:43]: This needs to be removed, right?

**Review by jessesuen — APPROVED** [review]
Conditional approval after removal of LIST_LIMIT from documentation.

File: workflow/controller/controller.go [hunk 3]
```diff
@@ -803,6 +803,7 @@ func (wfc *WorkflowController) tweakListOptions(options *metav1.ListOptions) {
 	labelSelector := labels.NewSelector().
 		Add(util.InstanceIDRequirement(wfc.Config.InstanceID))
 	options.LabelSelector = labelSelector.String()
+	options.Limit = int64(env.LookupEnvIntOr("LIST_LIMIT", 200))
```
Comment (jessesuen) [inline workflow/controller/controller.go:806]: Will this paginate properly? Or will it simply cut it off at 200

**Pull request merged**